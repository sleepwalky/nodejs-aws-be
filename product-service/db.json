[
  {
    "count": 4,
    "description": "Дополнение",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80aa",
    "price": 87,
    "img": "https://igraj.by/upload/iblock/f5a/f5a48e0762001ad9be2371423df702ab.jpg",
    "title": "Крылья: Птицы Европы"
  },
  {
    "count": 6,
    "description": "Short Product Description3",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a0",
    "price": 175,
    "img": "https://igraj.by/upload/Sh/imageCache/6ac/511/2f4c76754a96582d6db19f00e72a312e.jpg",
    "title": "Маракайбо"
  },
  {
    "count": 7,
    "description": "Short Product Description2",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a2",
    "price": 17,
    "img": "https://igraj.by/upload/Sh/imageCache/86d/efa/f6f4124a393a4c6c0a641c0625ae4368.jpg",
    "title": "Соображарий Супер"
  },
  {
    "count": 12,
    "description": "Short Product Description7",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a1",
    "price": 116,
    "img": "https://igraj.by/upload/Sh/imageCache/259/95a/420580760fcf1a89aac7c1929b2860f6.jpg",
    "title": "Цивилизация Сида Мейера: Новый рассвет. Терра инкогнита"
  },
  {
    "count": 7,
    "description": "Short Product Description2",
    "id": "7567ec4b-b10c-48c5-9345-fc73c48a80a3",
    "price": 31,
    "img": "https://igraj.by/upload/Sh/imageCache/47c/d51/cae2ade43f0046c6da31e6dac6c8840e.jpg",
    "title": "Бэнг! Звёздные Войны"
  },
  {
    "count": 8,
    "description": "Short Product Description4",
    "id": "7567ec4b-b10c-48c5-9345-fc73348a80a6",
    "price": 420,
    "img": "https://igraj.by/upload/Sh/imageCache/5cb/8df/3bb1ac23ac7ffa8255a6bff45d3fb400.jpg",
    "title": "Немезида / Nemesis"
  },
  {
    "count": 2,
    "description": "Short Product Descriptio1",
    "id": "7567ec4b-b10c-48c5-9445-fc73c48a80a5",
    "price": 175,
    "img": "https://igraj.by/upload/Sh/imageCache/633/662/aeb2601974000ce9a38f8446f233e766.jpg",
    "title": "Остров Духов"
  },
  {
    "count": 3,
    "description": "Short Product Description7",
    "id": "7567ec4b-b10c-45c5-9345-fc73c48a80a4",
    "price": 105,
    "img": "https://igraj.by/upload/Sh/imageCache/b1e/936/7108af77ff6beaf9c3042c8cf0a280e1.png",
    "title": "Ярость Дракулы"
  }
]
