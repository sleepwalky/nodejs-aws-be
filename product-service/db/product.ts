import { Client, ClientConfig } from 'pg';
import { Product } from '../models/Product';

const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;
const dbOptions: ClientConfig = {
  host: PG_HOST,
  port: parseInt(PG_PORT, 10),
  database: PG_DATABASE,
  user: PG_USERNAME,
  password: PG_PASSWORD,
  ssl: {
    rejectUnauthorized: false, // to avoid warring in this example
  },
  connectionTimeoutMillis: 5000, // time in millisecond for termination of the database query
};

export const getProductsFromDb = async () => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    const { rows: products } = await client.query(
      `select * from products p inner join stocks s on p.id = s.product_id `
    );
    console.log(products);
    return products;
  } catch (err) {
    console.error('Error during database request executing:', err);
    throw err;
  } finally {
    client.end();
  }
};

export const getProductByIdFromDb = async (id: string) => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    const { rows: products } = await client.query(
      `select * from products p inner join stocks s on p.id = s.product_id where p.id = '${id}'`
    );
    console.log(products);
    return products[0];
  } catch (err) {
    console.error('Error during database request executing:', err);
    throw err;
  } finally {
    client.end();
  }
};

export const createProductDb = async ({
  description,
  title,
  img,
  count,
  price,
}: Product) => {
  const client = new Client(dbOptions);
  await client.connect();

  try {
    await client.query('BEGIN');
    const queryText =
      'INSERT INTO products(title, description, price, img) VALUES($1, $2, $3, $4) RETURNING id';
    const insertProductValues = [title, description, price, img];
    console.log(insertProductValues);
    const res = await client.query(queryText, insertProductValues);
    console.log('succesfull insert into products');
    const insertStocksText =
      'INSERT INTO stocks(product_id, count) VALUES ($1, $2)';
    const insertStocksValues = [res.rows[0].id, count];
    await client.query(insertStocksText, insertStocksValues);
    await client.query('COMMIT');
    return await getProductByIdFromDb(res.rows[0].id);
  } catch (e) {
    await client.query('ROLLBACK');
    console.log(e);
    throw e;
  } finally {
    client.end();
  }
};
