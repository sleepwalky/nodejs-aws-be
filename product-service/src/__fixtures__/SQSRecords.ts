export const records = [
  {
    messageId: '7c9f1c6d-9713-40a9-988d-695c58eb2df2',
    receiptHandle:
      'AQEBhRJ6Xp2VRBRbMmp1J+njY4S2rs4PZ3WLO8Vg/bZGJ5uJVpIQxZoaEKv6CniLn4wmHcIvesPYolMrzOyOxq0h73ON/Z11XN4UEI2mt5yAkTD7ytKwZ65ah2LSW7JtW5WGr56n0QsxPPqWc91FHy8kajBj2d0Q4R/oAnyxFs4nHom/mQJ9sTQlEjL9ta8tkEha9yRpv1dhSbPQ3hnLxeAtsTvYXFWnnWXu/1VSLJKczdFEr9jsTwFpo42q7dQmTM1C+numSQUbIdzj3ZfrouYPCG5JFbgqEcbXl8zyZINa5JCgXOk7VsaneZz+zrqdkFsBBDzZYu4ca3kAoNh1Fzq/z7YTuJBb/vkGTUSDQGbqTXxE/A/rmdsVFczGDRX/i+DH7gAt+s1KMA60097dtF67gw==',
    body:
      '{"title":"Крылья: Птицы Европы","description":"Дополнение","price":"87","img":"https://igraj.by/upload/iblock/f5a/f5a48e0762001ad9be2371423df702ab.jpg","count":"4"}',
    attributes: [Object],
    messageAttributes: {},
    md5OfBody: '987e4b40074a6608533486cd68a4207a',
    eventSource: 'aws:sqs',
    eventSourceARN: 'arn:aws:sqs:eu-west-1:120641612187:catalog-items-queue',
    awsRegion: 'eu-west-1',
  },
  {
    messageId: 'ed37bfe0-fce0-4d0c-ae0d-a2995e83eb38',
    receiptHandle:
      'AQEBNcvHoQrU8Wxu8V/W7ClNkR2SA+yB9S8464g0vJnClJ4Zj/fpxg0wOI+/vDEkUXkAHATWGa9XrNm0z0dY/91g4kuNvIGI0X5tI75ZKEYkxz/VPevmw4a6boENw3Sl5j5GEM4H6lKzJX6KwQoAEZPLq0Ag6udimtQJch3DR970sMw0p5mBp2z0oBtlDPx1PS1wWwjY/FTy+cWYkT5TrqNQQCu5EdM6bN3gunlnWlNr898pEJHhpx/NTZyjiukXv9n1CRMgRjgrwL/Ud1tA5Lj5g4VMDxOup8ky8AriLq0DePAamRLaICcUlC5S0LL34q5rp0oW/m8MuA8jMAshoIsb2lyyzr+RObvODxHfuAc8EVE1BhKdYnIpTq6BHAVuu+0VbrAmpcaz6SP+L7Mll0JyKA==',
    body:
      '{"title":"Остров Духов","description":"Остров духов описание","price":"175","img":"https://igraj.by/upload/Sh/imageCache/633/662/aeb2601974000ce9a38f8446f233e766.jpg","count":"2"}',
    attributes: [Object],
    messageAttributes: {},
    md5OfBody: '34a2f8d666abf5a8ae82b6663f6b595f',
    eventSource: 'aws:sqs',
    eventSourceARN: 'arn:aws:sqs:eu-west-1:120641612187:catalog-items-queue',
    awsRegion: 'eu-west-1',
  },
  {
    messageId: '4a639990-3b09-4758-993f-a17411ac084a',
    receiptHandle:
      'AQEBaev+MsRQ8nlbYUQkrawLJdvQlX1CPDOQfUPvHCzxaE/RISAFuYaj4Zs8G7HvzfOE8zSntD39/QShAg8Lb8sbtHDKJqO+KOQIAB2kQvx7ziSEWWleBfcdjnQXQANlUXEPxMfKbo9PAhUQ1spqHtNU+/t1SADqPuD/zp286n22cS/DfpikCuHNjocCOBAfAA0kS6ZhPBWgTGNjWqRREL66BAN65n+u8QdMyXA4jT4K/VvLR1CpFe15kaqXwCPa/Ug9D12JPgTgJvWWOLR/O+9ZXWlMvYxANadgv+xIoyiKYuBc4xJKWsEumWrNDnsl+NQo9WHm9JzGNPNwWdDuDcX4/l0KGv5r0V9lIpNvRwokZMyLmZ2Ybd3Pnx3SAwlU6tvP7Pqq2Dx67KFZLQEp3f2/Ow==',
    body:
      '{"title":"Бэнг! Звёздные Войны","description":"Бэнг Описание","price":"31","img":"https://igraj.by/upload/Sh/imageCache/47c/d51/cae2ade43f0046c6da31e6dac6c8840e.jpg","count":"7"}',
    attributes: [Object],
    messageAttributes: {},
    md5OfBody: 'c1d39a01a13b9739824989da742b1103',
    eventSource: 'aws:sqs',
    eventSourceARN: 'arn:aws:sqs:eu-west-1:120641612187:catalog-items-queue',
    awsRegion: 'eu-west-1',
  },
];
