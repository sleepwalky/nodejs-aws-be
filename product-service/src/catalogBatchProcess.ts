import { SQSHandler } from 'aws-lambda';
import 'source-map-support/register';

export const catalogBatchProcess: SQSHandler = (event) => {
  console.log(event);

  const { Records } = event;

  for (const record of Records) {
    console.log(record);
  }
};
