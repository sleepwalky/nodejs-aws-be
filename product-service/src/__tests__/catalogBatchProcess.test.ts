import { catalogBatchProcess } from '../catalogBatchProcess';
import { SNS } from 'aws-sdk/';
import { records } from '../__fixtures__/SQSRecords';
import { createProductDb } from '../../db/product';

jest.mock('../../db/product');

// jest.mock('aws-sdk', () => {
//   return jest.fn().mockImplementation(() => {
//     return {
//       SNS: {
//         publish: jest.fn(),
//       },
//     };
//   });
// });

// jest.mock('aws-sdk', () => ({
//   SNS: jest.fn().mockImplementation(() => ({
//     publish: jest.fn().mockImplementation(() => ({
//       promise: jest.fn(),
//     })),
//   })),
// }));

jest.mock('aws-sdk', () => ({
  SNS: mockClass({
    publish: jest.fn(),
  }),
}));

describe('catalogBatchProcess', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should process records', async () => {
    (SNS.prototype.publish as jest.Mock).mockImplementation(() => ({
      promise: jest.fn().mockResolvedValue(undefined),
    }));

    await catalogBatchProcess(
      {
        Records: records,
      } as any,
      {} as any,
      () => {}
    );

    expect(SNS.prototype.publish).toHaveBeenCalledTimes(records.length);
  });

  it('should publish success messages', async () => {
    (SNS.prototype.publish as jest.Mock).mockImplementation(() => ({
      promise: jest.fn().mockResolvedValue(undefined),
    }));

    await catalogBatchProcess(
      {
        Records: records,
      } as any,
      {} as any,
      () => {}
    );

    expect(SNS.prototype.publish).toHaveBeenCalledTimes(records.length);
  });

  it('should publish error messages', async () => {
    (SNS.prototype.publish as jest.Mock).mockImplementation(() => ({
      promise: jest.fn().mockResolvedValue(undefined),
    }));
    (createProductDb as jest.Mock).mockRejectedValue('some error');

    await catalogBatchProcess(
      {
        Records: [records[0]],
      } as any,
      {} as any,
      () => {}
    );

    expect(
      (SNS.prototype.publish as jest.Mock).mock.calls[0][0]
    ).toMatchSnapshot();
  });
});

function mockClass(spec) {
  const proto = jest.fn(() => spec);
  proto.prototype = spec;
  return proto;
}
