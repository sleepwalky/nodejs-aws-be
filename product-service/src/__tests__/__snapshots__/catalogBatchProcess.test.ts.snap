// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalogBatchProcess should publish error messages 1`] = `
Object {
  "Message": "
          There was an error:
          {\\"title\\":\\"Крылья: Птицы Европы\\",\\"description\\":\\"Дополнение\\",\\"price\\":\\"87\\",\\"img\\":\\"https://igraj.by/upload/iblock/f5a/f5a48e0762001ad9be2371423df702ab.jpg\\",\\"count\\":\\"4\\"}
          \\"some error\\"
          ",
  "MessageAttributes": Object {
    "name": Object {
      "DataType": "String",
      "StringValue": "bar",
    },
  },
  "TopicArn": undefined,
}
`;
